ocaml/typing/jkind.ml

```diff
@@ -261,6 +274,238 @@ end
+    | External (* not managed by the garbage collector *)
+    | External64 (* not managed by the garbage collector on 64-bit systems *)
+    | Internal (* managed by the garbage collector *)
+
+  (* CR layouts v2.8: Either use this or remove it *)
+  let _to_string = function
+    | External -> "external"
+    | External64 -> "external64"
+    | Internal -> "internal"
+
+  let max = Internal
+
+  let min = External
+
+  let equal e1 e2 =
+    match e1, e2 with
+    | External, External -> true
+    | External64, External64 -> true
+    | Internal, Internal -> true
+    | (External | External64 | Internal), _ -> false
+
+  let sub t1 t2 : Sub_result.t =
+    match t1, t2 with
+    | External, External -> Equal
+    | External, (External64 | Internal) -> Sub
+    | External64, External -> Not_sub
+    | External64, External64 -> Equal
+    | External64, Internal -> Sub
+    | Internal, (External | External64) -> Not_sub
+    | Internal, Internal -> Equal
+
+  let intersection t1 t2 =
+    match t1, t2 with
+    | External, (External | External64 | Internal)
+    | (External64 | Internal), External ->
+      External
+    | External64, (External64 | Internal) | Internal, External64 -> External64
+    | Internal, Internal -> Internal
+
+  module Debug_printers = struct
+    open Format
+
+    let t ppf = function
+      | External -> fprintf ppf "External"
+      | External64 -> fprintf ppf "External64"
+      | Internal -> fprintf ppf "Internal"
+  end
+end
+
+module Const = struct
+  type t =
+    { layout : Layout.Const.t;
+      externality : Externality.t
+    }
+
+  (* CR layouts v2.8: remove this *)
+  let to_legacy_jkind { layout; externality } : Jane_asttypes.const_jkind =
+    match layout, externality with
+    | Any, _ -> Any
+    | Sort Value, Internal -> Value
+    | Sort Value, External64 -> Immediate64
+    | Sort Value, External -> Immediate
+    | Sort Void, _ -> Void
+    | Sort Float64, _ -> Float64
+
+  (* CR layouts v2.8: do a better job here *)
+  let to_string t = const_jkind_to_string (to_legacy_jkind t)
+
+  let sub { layout = lay1; externality = ext1 }
+      { layout = lay2; externality = ext2 } =
+    Sub_result.combine (Layout.Const.sub lay1 lay2) (Externality.sub ext1 ext2)
+end
+
+module Desc = struct
+  type t =
+    | Const of Const.t
+    | Var of Sort.var (* all modes will be [max] *)
+
+  let format ppf =
+    let open Format in
+    function
+    | Const c -> fprintf ppf "%s" (Const.to_string c)
+    | Var v -> fprintf ppf "%s" (Sort.var_name v)
+
+  (* considers sort variables < Any, but otherwise just checks for equality.
+     Never does mutation.
+     Pre-condition: no filled-in sort variables. *)
+  let sub d1 d2 : Sub_result.t =
+    match d1, d2 with
+    | Const c1, Const c2 -> Const.sub c1 c2
+    | Var _, Const { layout = Any; externality = Internal } -> Sub
+    | Var v1, Var v2 -> if v1 == v2 then Equal else Not_sub
+    | Const _, Var _ | Var _, Const _ -> Not_sub
+end
+
+module Jkind_ = struct
```

Comment:
Why `Jkind_` rather than `Jkind`?

Reply:
Because I didn't want `Jkind.t` to mean one thing inside this file and a different thing everywhere else. I thought of `Jkind_internal` or something like that, but I didn't like that much more either. Open to suggestions.

Reply:
Is the long term plan to keep this module, or to one day combine the two types (so the history is just another field of what is currently called `Jkind_.t`?  (And we're not doing that today to avoid annoying merge conflicts with the errors work).

If so, I think this doesn't matter so much.  And I think that would be a fine design (maybe better - fewer boxes to allocate and pointers to follow).

If not, I prefer a name like `Jkind_internal` or just `Internal`.  Another option would be to name this module `Jkind_desc` or the type `jkind_desc`, by analogy with other parts of the type system.

Reply:
My thought was to keep this separation indefinitely. It allows us to talk about the meaty parts of a jkind, without regard to the history. Though I agree it's unfortunately a little allocate-y. If you think it's worthwhile, I can make a ticket about possibly eliminiating the internal type later.

I'm quite happy with `Jkind_desc` -- that's definitely better than what I've done. I will make that change.